nixos/modules/services/backup/postgresql-wal-receiver.nix

```diff
@@ -0,0 +1,210 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  user = "postgresql-wal-receiver";
+  group = "postgresql-wal-receiver";
+
+  receiverSubmodule = {
+    options = {
+      postgresqlPackage = mkOption {
+        type = types.package;
+        example = literalExample "pkgs.postgresql_9_6";
+        description = ''
+          PostgreSQL package to use.
+        '';
+      };
+
+      directory = mkOption {
+        type = types.path;
+        example = literalExample "/mnt/pg_wal/main/";
+        description = ''
+          Directory to write the output to.
+        '';
+      };
+
+      noSync = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          This option causes <command>pg_receivewal</command> to not force WAL data to be flushed to disk.
+          This is faster, but means that a subsequent operating system crash can leave the WAL segments corrupt.
+          Generally, this option is useful for testing but should not be used when doing WAL archiving on a production deployment.
+
+          This option requires PostgreSQL >= 11 and is incompatible with <option>synchronous</option>.
+        '';
+      };
+
+      statusInterval = mkOption {
+        type = types.int;
+        default = 10;
+        description = ''
+          Specifies the number of seconds between status packets sent back to the server.
+          This allows for easier monitoring of the progress from server.
+          A value of zero disables the periodic status updates completely,
+          although an update will still be sent when requested by the server, to avoid timeout disconnect.
+        '';
+      };
+
+      slot = mkOption {
+        type = types.str;
+        default = "";
+        example = "some_slot_name";
+        description = ''
+          Require <command>pg_receivewal</command> to use an existing replication slot (see
+          <link xlink:href="https://www.postgresql.org/docs/current/warm-standby.html#STREAMING-REPLICATION-SLOTS">Section 26.2.6 of the PostgreSQL manual</link>).
+          When this option is used, <command>pg_receivewal</command> will report a flush position to the server,
+          indicating when each segment has been synchronized to disk so that the server can remove that segment if it is not otherwise needed.
+
+          When the replication client of <command>pg_receivewal</command> is configured on the server as a synchronous standby,
+          then using a replication slot will report the flush position to the server, but only when a WAL file is closed.
+          Therefore, that configuration will cause transactions on the primary to wait for a long time and effectively not work satisfactorily.
+          The option <option>synchronous</option> must be specified in addition to make this work correctly.
+        '';
+      };
+
+      synchronous = mkOption {
```

Comment:
I had to look into [source](https://git.postgresql.org/gitweb/?p=postgresql.git;a=blobdiff;f=src/bin/pg_basebackup/pg_receivewal.c;h=d801ea07fc8e421f88a0de31ee400fb5b7d880d0;hp=888ae6c571902be4c9f91fec1587aee8acafc87c;hb=5f3971291fc231bb65a38198b1bcb1c29ef63108;hpb=b7f3eb31405f1dbbf086e5a8f88569a6dc85157a) to understand why `noSync != !synchronous`

What do you think about making tristate option?
```
synchronous = mkOption {
  type = types.enum [ "synchronous" "fsync" "no-sync" ];
  default = "synchronous";
  description = ''
    Whether WAL is received synchronous with WAL generation. Possible options:

    </para><para><ul>
    <li><literal>"synchronous"</literal>: flush the WAL data to disk immediately after it has been received,
          and send a status packet back to server immediately after flushing, regardless of
          <option>statusInterval</option>. Requires server configured in synchronous standby mode</li>
    <li><literal>"fsync"</literal>: just flush the WAL data to disk immediately after it has been received</li>
    <li><literal>"no-sync"</literal>: do not even flush the WAL data to disk. This is fastest, but unsafe: 
          operating system crash can leave the WAL segments corrupt (requires PostgreSQL 11+).</li>
    </ul>
  '';
}
```

Reply:
I've removed the `noSync` option as it was meant for testing only.
The user can still use it through the added `extraArgs` option.